src/index.js

```diff
@@ -180,26 +223,62 @@ class Bitswap {
       blockListeners[cidStr] = (block) => {
         this.wm.cancelWants([cid])
         cleanupListener(cid)
-        callback(null, block)
+        retrieved.push(block)
+
+        if (retrieved.length === missing.length) {
+          finish(callback)
+        }
       }
 
-      this.notifications.once(unwantEvent, unwantListeners[cidStr])
-      this.notifications.once(blockEvent, blockListeners[cidStr])
+      this.notifications.once(
+        unwantEvent(cidStr),
+        unwantListeners[cidStr]
+      )
+      this.notifications.once(
+        blockEvent(cidStr),
+        blockListeners[cidStr]
+      )
     }
```

Comment:
- [ ] Q: Can we make this notification/tracking gets system be an external and tested module?